3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,253 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import './polyfills';
+import {dev} from '../src/log';
+import {initLogConstructor, setReportError} from '../src/log';
+import {IframeMessagingClient} from './iframe-messaging-client';
+import {AMP_ANALYTICS_3P_MESSAGE_TYPE} from '../src/3p-analytics-common';
+initLogConstructor();
+// TODO(alanorozco): Refactor src/error.reportError so it does not contain big
+// transitive dependencies and can be included here.
+setReportError(() => {});
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/**
+ * Receives messages bound for this cross-domain iframe, from all creatives
+ */
+class AmpAnalytics3pMessageRouter {
+  /** @param {!Window} win */
+  constructor(win) {
+
+    /** @private {!Window} */
+    this.win_ = win;
+
+    try {
+      const frameNameData = JSON.parse(this.win_.name);
+      if (!frameNameData.sentinel) {
+        const ex = {message: 'JSON parsed, but did not include sentinel.'};
+        throw ex;
+      }
+      // The sentinel is required by iframe-messaging-client, and is used to
+      // uniquely identify the frame as part of message routing
+      /** @const {string} */
+      this.sentinel_ = frameNameData.sentinel;
+    } catch (e) {
+      dev().warn(TAG_, 'Unable to set sentinel from name attr: ' +
+        this.win_.name + ': ' + e.message);
+      return;
+    }
+
+    /** @private {!Object<string, !AmpAnalytics3pCreativeMessageRouter>} */
+    this.creativeMessageRouters_ = {};
+
+    /**
+     * Handles communication between frames
+     * @private {!IframeMessagingClient}
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(this.sentinel_);
+    this.iframeMessagingClient_.registerCallback(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVES,
+      envelope => {
```

Comment:
I meant it in a similar sense to its use in https://docs.oracle.com/cd/E19509-01/820-4408/gguuw/index.html but somewhat less formal - it's just what I always use to name something whose sole purpose is to contain other thing(s), especially in transit (since that's the purpose of a real-world envelope). Happy to change if you prefer.